cpp/src/lance/arrow/dataset.cc

```diff
@@ -153,15 +161,31 @@ std::string LanceDataset::Impl::data_dir() const { return fs::path(base_uri) / k
 
 std::string LanceDataset::Impl::versions_dir() const { return fs::path(base_uri) / kVersionsDir; }
 
-::arrow::Result<std::unique_ptr<LanceDataset::Impl>> LanceDataset::Impl::WriteNewVersion(
-    std::shared_ptr<lance::format::Manifest> new_manifest) const {
+namespace {
+
+::arrow::Status WriteManifest(const std::shared_ptr<::arrow::fs::FileSystem>& fs,
+                              const std::string& base_uri,
+                              const std::shared_ptr<format::Manifest>& manifest) {
+  // Write the manifest version file.
+  // It only supports single writer at the moment.
+  auto version_dir = (fs::path(base_uri) / kVersionsDir).string();
+  ARROW_RETURN_NOT_OK(fs->CreateDir(version_dir));
   auto manifest_path = GetManifestPath(base_uri, manifest->version());
   {
+    // Keep the serialized time, which is when this particular version of dataset is visible.
```

Comment:
when you say "visible", you're just referring to the call to `fs->CreateDir`? At this point there's no data that's available right? Should the time stamp be when this version's data is available?

it's possible that creation of a new version takes some time. If someone else requests a new version while we're in the middle of creating a new version, what would happen?

Reply:
This is the last step of writing dataset, so at this point, the data files have already been written.

> it's possible that creation of a new version takes some time. If someone else requests a new version while we're in the middle of creating a new version, what would happen?

It does not allow multi-writer now

https://github.com/eto-ai/lance/pull/343/files/5d52e39c469095d48b413d144301c6009da36875#diff-b640faff51137caba09899ac37fb468dd1dbe1cc5e333ce2bb4dd8c43e53a12cR169-R170